consensus/src/units/mod.rs

```diff
@@ -200,34 +181,77 @@ pub(crate) type UncheckedSignedUnit<H, D, S> = UncheckedSigned<FullUnit<H, D>, S
 
 pub(crate) type SignedUnit<H, D, K> = Signed<FullUnit<H, D>, K>;
 
-#[derive(Clone, Eq, PartialEq, Debug, Decode, Encode)]
-pub struct Unit<H: Hasher> {
-    pre_unit: PreUnit<H>,
-    hash: H::Hash,
-}
+/// Abstract representation of a unit from the Dag point of view.
+pub trait Unit: 'static + Send + Clone {
+    type Hasher: Hasher;
+
+    fn hash(&self) -> <Self::Hasher as Hasher>::Hash;
+
+    fn coord(&self) -> UnitCoord;
 
-impl<H: Hasher> Unit<H> {
-    pub(crate) fn new(pre_unit: PreUnit<H>, hash: H::Hash) -> Self {
-        Unit { pre_unit, hash }
+    fn control_hash(&self) -> &ControlHash<Self::Hasher>;
+
+    fn creator(&self) -> NodeIndex {
+        self.coord().creator()
     }
-    pub(crate) fn creator(&self) -> NodeIndex {
-        self.pre_unit.creator()
+
+    fn round(&self) -> Round {
+        self.coord().round()
     }
-    pub(crate) fn round(&self) -> Round {
-        self.pre_unit.round()
+}
+
+pub trait WrappedUnit<H: Hasher>: Unit<Hasher = H> {
```

Comment:
I think I understand the purpose of `Unit` - you want the store be abstract of the unit type. But I'm not sure I get the purpose of `WrappedUnit`.

Reply:
In `validation` we get a store with units of a different type (with parents in practice, but there is no reason for `validation` to know that) as an argument, but we have to be able to use these as the type we have there, i.e. `SignedUnit`. It would just as well be an `Into<SignedUnit<H, D, MK>> + Unit<H>` I guess?

Reply:
Aight, we can keep this simplification in the back of our heads. I won't ask you to change it to `Into<....`, unless you think it's actually more convenient.